### PR TITLE
Performance improvements pt. 1

### PR DESCRIPTION
**Pull request by ajafff**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [x] ~~New feature, bugfix, or~~ enhancement
  - [ ] Includes tests
- [ ] Documentation update
- [x] Enable CircleCI for your fork (https://circleci.com/add-projects)

#### What changes did you make?

This is a first attempt to improve performance of core rules. It's basically #1747 ... search for `\.get((Full)?Text|Start|(LeadingTrivia)?Width|(First|Last)Token|Child(ren|Count))\(\)|\.getChildAt\(\d+\)` and pass SourceFile as parameter.

To keep the diff somewhat small, I chose to submit the changes in chunks as I progress. So here we go from `alignRule` until `noDefaultExportRule`.

Of course this is only the low hanging fruit. When I find the time, I will look at the rules in more detail.

### CODE REVIEW

**Comment by ajafff** [comment]
As a side note: Maybe we could create a tslint rule for that? Maybe in a separate project as this would not be relevant for end users of tslint?

File: src/language/utils.ts [hunk 1]
```diff
@@ -81,9 +86,12 @@ export function hasModifier(modifiers: ts.ModifiersArray | undefined, ...modifie
         return false;
     }
 
-    return modifiers.some((m) => {
-        return modifierKinds.some((k) => m.kind === k);
-    });
+    for (const modifier of modifiers) {
```
Comment (nchen63) [inline src/language/utils.ts]: why was this changed?

File: src/language/utils.ts [hunk 2]
```diff
@@ -54,11 +54,16 @@ export function createCompilerOptions(): ts.CompilerOptions {
 }
 
 export function doesIntersect(failure: RuleFailure, disabledIntervals: IDisabledInterval[]) {
-    return disabledIntervals.some((interval) => {
-        const maxStart = Math.max(interval.startPosition, failure.getStartPosition().getPosition());
-        const minEnd = Math.min(interval.endPosition, failure.getEndPosition().getPosition());
-        return maxStart <= minEnd;
-    });
+    const failureStart = failure.getStartPosition().getPosition();
```
Comment (nchen63) [inline src/language/utils.ts]: why change?

File: src/language/utils.ts [hunk 3]
```diff
@@ -117,22 +125,17 @@ export function getBindingElementVariableDeclaration(node: ts.BindingElement): t
...
-    return undefined;
-}
-
 /**
  * Finds a child of a given node with a given kind.
  * Note: This uses `node.getChildren()`, which does extra parsing work to include tokens.
  */
-export function childOfKind(node: ts.Node, kind: ts.SyntaxKind): ts.Node | undefined {
-    return find(node.getChildren(), (child) => child.kind === kind);
+export function childOfKind(node: ts.Node, kind: ts.SyntaxKind, sourceFile?: ts.SourceFile): ts.Node | undefined {
```
Comment (nchen63) [inline src/language/utils.ts]: why change?

**Comment by nchen63** [comment]
I'm wary of making perf improvements without a performance profiler indicating a problem or a benchmark improvement.

**Comment by ericanderson** [comment]
I can't speak to whether or not there is a performance issue that __needs__ to be fixed, but for loops seem to nearly always be faster than method invocations.

A simple test: http://jsperf.com/array-some-vs-loop

**Comment by ajafff** [comment]
@ericanderson you're right at the moment, but eventually there will be no difference, hopefully.

However, that specific commit was not intended to be included in this or any other PR, so I reverted it.

**Comment by ajafff** [comment]
@nchen63 You're right, I should have done the profiling before changing everything.
I just profiled some pretty big libraries including `TypeScript` with only `one-line`-Rule enabled, which has the most usages of `node.getChildren` and friends. The result was crushing this PR:
Out of 24s runtime only 16ms were spent in `getSourceFileOfNode`.

I'll close this PR and maybe you can even close #1747.
